metricbeat/module/system/diskio/diskstat_linux.go

```diff
@@ -0,0 +1,96 @@
+// +build linux
+
+package diskio
+
+import (
+	"github.com/elastic/beats/metricbeat/module/system"
+	"github.com/pkg/errors"
+	"github.com/shirou/gopsutil/disk"
+)
+
+func Get_CLK_TCK() uint32 {
+	//return uint32(C.sysconf(C._SC_CLK_TCK))
+	//NOTE: _SC_CLK_TCK should be fetched from sysconf using cgo
+	return uint32(100)
+}
+
+func NewDiskIOStat() *DiskIOStat {
+	d := &DiskIOStat{}
+	d.lastDiskIOCounters = make(map[string]disk.IOCountersStat)
+	return d
+}
+
+// create current cpu sampling
+// need call as soon as get IOCounters
+func (stat *DiskIOStat) OpenSampling() error {
+	return stat.curCpu.Get()
+}
+
+func (stat *DiskIOStat) CalIOStatistics(counter disk.IOCountersStat) (DiskIOMetric, error) {
+	var last disk.IOCountersStat
+	var ok bool
+	var result DiskIOMetric
+
+	// if last counter not found, create one and return all 0
+	if last, ok = stat.lastDiskIOCounters[counter.Name]; !ok {
+		stat.lastDiskIOCounters[counter.Name] = counter
+		return result, nil
+	}
+
+	// calculate the delta ms between the CloseSampling and OpenSampling
+	deltams := 1000.0 * float64(stat.curCpu.Total()-stat.lastCpu.Total()) / float64(system.NumCPU) / float64(Get_CLK_TCK())
+	if deltams <= 0 {
+		return result, errors.New("The delta cpu time between close sampling and open sampling is less or equal to 0")
+	}
```

Comment:
i think these metrics calculation is not easy on the fly, so calculation and report to backend directly is necessary here